lib/IRGen/IRGenSIL.cpp

```diff
@@ -1226,13 +1226,19 @@ IRGenSILFunction::IRGenSILFunction(IRGenModule &IGM, SILFunction *f)
   }
   if (IGM.IRGen.Opts.Sanitizers & SanitizerKind::Thread) {
     auto declContext = f->getDeclContext();
-    if (declContext && isa<DestructorDecl>(declContext))
+    if (f->getLoweredFunctionType()->isCoroutine()) {
+      // Disable TSan in coroutines; the instrumentation currently interferes
+      // with coroutine structural invariants.
+    } else if (declContext && isa<DestructorDecl>(declContext)) {
       // Do not report races in deinit and anything called from it
       // because TSan does not observe synchronization between retain
       // count dropping to '0' and the object deinitialization.
+      // Also don't report races in coroutines because the instrumentation
```

Comment:
I think this comment "Also don't report races in coroutines..." is misplaced since we're not using "sanitize_thread_no_checking_at_run_time" for coroutines. Can we just remove the comment?

Reply:
I will upload a new version of the patch after this one (presumably) passes the full test suite.